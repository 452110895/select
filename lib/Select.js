/** @jsx React.DOM */

/**
 * Select
 */
var React = require('react');
var rcUtil = require('rc-util');
var joinClasses = rcUtil.joinClasses;
var classSet = rcUtil.classSet;
var KeyCode = rcUtil.KeyCode;
var Menu = require('rc-menu');
var MenuItem = Menu.Item;

function isMultipleOrTags(props) {
  return props.multiple || props.tags;
}

function noop() {
}

function getValueFromOptionChild(child) {
  var optionProps = child.props;
  var children = optionProps.children;
  var ret;
  if (optionProps.value !== undefined) {
    if (typeof optionProps.value === 'string') {
      ret = optionProps.value;
    }
  } else if (typeof children === 'string') {
    ret = children;
  }
  if (!ret && !child.props.disabled) {
    throw new Error('must set value string on Option element!');
  }

  return ret;
}

function getPropsFromOption(child) {
  var ret = {
    key: getValueFromOptionChild(child),
    value: getValueFromOptionChild(child)
  };
  var optionProps = child.props;
  for (var i in optionProps) {
    if (i !== 'children') {
      ret[i] = optionProps[i];
    }
  }
  return ret;
}

function normValue(value) {
  if (value === undefined) {
    value = [];
  } else if (!Array.isArray(value)) {
    value = [value];
  }
  return value;
}

class Select extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      value: normValue(props.value),
      inputValue: ''
    };
    ['handleClick',
      'handleKeyDown',
      'handleInputKeyDown',
      'handleInputChange',
      'handleFocus',
      'handleBlur',
      'handleClearSelection',
      'handleMenuSelect',
      'handleMenuDeselect'].forEach((m)=> {
        this[m] = this[m].bind(this);
      });
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      value: normValue(nextProps.value)
    });
  }

  renderFilterOptions() {
    var inputValue = this.state.inputValue;
    var sel = [];
    var props = this.props;
    var childrenKeys = [];
    var filterOption = props.filterOption;
    var tags = props.tags;
    React.Children.forEach(props.children, (child)=> {
      if (!filterOption || !inputValue || !child.props.disabled && getValueFromOptionChild(child).indexOf(inputValue) > -1) {
        sel.push(child);
      }
      if (tags && !child.props.disabled) {
        childrenKeys.push(getValueFromOptionChild(child));
      }
    });
    if (tags) {
      var value = this.state.value || [];
      value = value.filter((v)=> {
        return childrenKeys.indexOf(v) === -1 && (!inputValue || v.indexOf(inputValue) > -1);
      });
      sel = sel.concat(value.map((v)=> {
        return <MenuItem value={v}>{v}</MenuItem>;
      }));
      if (inputValue) {
        var notFindInputItem = sel.every((s)=> {
          return getValueFromOptionChild(s) !== inputValue;
        });
        if (notFindInputItem) {
          sel.unshift(<MenuItem value={inputValue}>{inputValue}</MenuItem>);
        }
      }
    }
    if (!sel.length) {
      sel = <MenuItem disabled value='NOT_FOUND'>{props.notFoundContent}</MenuItem>;
    }
    return sel;
  }

  setOpenState(open) {
    this.setState({
      open: open
    }, ()=> {
      if (open || isMultipleOrTags(this.props) || this.props.combobox) {
        this.refs.input.getDOMNode().focus();
      } else {
        this.refs.selection.getDOMNode().focus();
      }
    });
  }

  handleInputChange(e) {
    var val = e.target.value;
    this.setState({
      inputValue: val,
      open: true
    });
    if (this.props.combobox) {
      this.setState({
        value: val ? [val] : []
      });
      this.props.onChange(val);
    }
  }

  handleClick() {
    this.setOpenState(!this.state.open);
  }

  // combobox ignore
  handleKeyDown(e) {
    var keyCode = e.keyCode;
    if (keyCode === KeyCode.ENTER || e.keyCode === KeyCode.DOWN) {
      this.handleClick(e);
      e.preventDefault();
    }
  }

  handleInputKeyDown(e) {
    if (e.keyCode === KeyCode.DOWN) {
      if (!this.state.open) {
        this.setState({
          open: true
        });
        e.preventDefault();
        e.stopPropagation();
        return;
      }
    } else if (e.keyCode === KeyCode.ESC) {
      if (this.state.open) {
        this.setOpenState(false);
        e.preventDefault();
        e.stopPropagation();
      }
      return;
    }
    if (this.refs.menu) {
      if (this.refs.menu.handleKeyDown(e)) {
        e.preventDefault();
        e.stopPropagation();
      }
    }
  }

  handleMenuSelect(key, item) {
    var value;
    var props = this.props;
    var selectedValue = item.props.value;
    if (isMultipleOrTags(props)) {
      value = this.state.value.concat();
      value.push(selectedValue);
    } else {
      if (this.state.value[0] === selectedValue) {
        this.setOpenState(false);
        return;
      }
      value = [selectedValue];
    }
    props.onSelect(selectedValue);
    props.onChange(isMultipleOrTags(props) ? value : value[0]);
    this.setState({
      value: value,
      inputValue: ''
    });
    this.setOpenState(false);
    if (props.combobox) {
      this.setState({
        inputValue: value[0]
      });
    }
  }

  handleMenuDeselect(key, item, e) {
    if (e.type === 'click') {
      this.removeSelected(item.props.value);
    }
    this.setState({
      inputValue: ''
    });
    this.setOpenState(false);
  }

  handleBlur() {
    if (this._blurTimer) {
      clearTimeout(this._blurTimer);
    }
    this._blurTimer = setTimeout(() => {
      this.setState({
        open: false
      });
    }, 100);
  }

  handleFocus() {
    if (this._blurTimer) {
      clearTimeout(this._blurTimer);
    }
  }

  removeSelected(selectedValue) {
    var props = this.props;
    var value = this.state.value.filter((v)=> {
      return v !== selectedValue;
    });
    var canMultiple = isMultipleOrTags(props);
    if (canMultiple) {
      props.onDeselect(selectedValue);
    }
    props.onChange(isMultipleOrTags(props) ? value : value[0]);
    this.setState({
      value: value
    });
  }

  handleClearSelection(e) {
    e.stopPropagation();
    if (this.state.inputValue || this.state.value.length) {
      this.props.onChange(isMultipleOrTags(this.props) ? [] : undefined);
      this.setState({
        value: [],
        inputValue: ''
      });
    }
    this.setOpenState(false);
  }

  renderMenu(children) {
    var props = this.props;
    var menuProps = {};
    if (isMultipleOrTags(props)) {
      menuProps.onDeselect = this.handleMenuDeselect;
    }
    var value = this.state.value;
    var menuItems = React.Children.map(children, this.renderOption);
    var selectedKeys = [];
    React.Children.forEach(menuItems, (item) => {
      if (value.indexOf(item.props.value) !== -1) {
        selectedKeys.push(item.key);
      }
    });
    var activeKey;
    if (selectedKeys.length === 1) {
      activeKey = selectedKeys[0];
    }
    return <Menu
      ref="menu"
      onSelect={this.handleMenuSelect}
      activeFirst={true}
      activeKey={activeKey}
      multiple={isMultipleOrTags(props)}
      focusable={false}
      {...menuProps}
      selectedKeys={selectedKeys}
      prefixCls={props.prefixCls + '-menu'}>
          {menuItems}
    </Menu>;
  }

  renderTopControlNode(input) {
    var value = this.state.value;
    var prefixCls = this.props.prefixCls;
    var allowClear = this.props.allowClear;
    var clear = <span className={prefixCls + '-selection__clear'}
      onClick={this.handleClearSelection}>×</span>;
    var props = this.props;
    // single and not combobox, input is inside dropdown
    if (!props.combobox && !isMultipleOrTags(props)) {
      return <span className={prefixCls + '-selection__rendered'}>
          {value[0]}
          {allowClear ? clear : null}
      </span>;
    }
    var selectedValueNodes;
    if (isMultipleOrTags(props)) {
      selectedValueNodes = value.map((v) => {
        return (
          <li className={prefixCls + '-selection__choice'}>
            <span className={prefixCls + '-selection__choice__remove'}
              onClick={this.removeSelected.bind(this, v)}
            >×</span>
           {v}
          </li>
        );
      });
    }
    return (
      <ul className={prefixCls + '-selection__rendered'}>
          {selectedValueNodes}
          {allowClear && !isMultipleOrTags(props) ? clear : null}
        <li className={joinClasses(prefixCls + '-search', prefixCls + '-search--inline')}>{input}</li>
      </ul>
    );
  }

  renderRoot(children) {
    var props = this.props;
    var prefixCls = props.prefixCls;
    var rootCls = {};
    rootCls[prefixCls] = true;
    if (this.state.open) {
      rootCls[prefixCls + '-open'] = true;
    }
    return (
      <span className={joinClasses(props.className, classSet(rootCls))} dir="ltr"
        onFocus={this.handleFocus}
        onBlur={this.handleBlur}>
        {children}
      </span>
    );
  }

  renderOption(child) {
    var props = getPropsFromOption(child);
    return <MenuItem {...props}>{child.props.children}</MenuItem>;
  }

  render() {
    var props = this.props;
    var multiple = isMultipleOrTags(props);
    var state = this.state;
    var prefixCls = props.prefixCls;

    var input = (
      <input ref="input"
        onChange={this.handleInputChange}
        onKeyDown={this.handleInputKeyDown}
        value={state.inputValue}
        className={prefixCls + '-search__field'}
        role="textbox" />
    );

    var children = this.renderFilterOptions();
    var ctrlNode = this.renderTopControlNode(input);
    var dropDown;
    if (state.open) {
      // single and not combobox, input is inside dropdown
      dropDown = <span key="dropdown" className= {joinClasses(prefixCls + '-dropdown', prefixCls + '-dropdown--below')} tabIndex="-1">
      {multiple || props.combobox ? null : <span className={joinClasses(prefixCls + '-search', prefixCls + '-search--dropdown')}>{input}</span>}
        {this.renderMenu(children)}
      </span>;
    }

    var extraSelectionProps = {};
    if (!props.combobox) {
      extraSelectionProps = {
        onKeyDown: this.handleKeyDown,
        tabIndex: 0
      };
    }

    return this.renderRoot([
      <span ref="selection"
        key="selection"
        className={joinClasses(prefixCls + '-selection',
          prefixCls + '-selection--' + (multiple ? 'multiple' : 'single'))}
        role="combobox"
        aria-autocomplete="list"
        onClick={this.handleClick}
        aria-haspopup="true"
        aria-expanded={state.open}
      {...extraSelectionProps}
      >
        {ctrlNode}
        {multiple ? null :
          <span key="arrow" className={prefixCls + '-arrow'}>
            <b></b>
          </span>}
      </span>,
      dropDown
    ]);
  }
}

Select.propTypes = {
  multiple: React.PropTypes.bool,
  tags: React.PropTypes.bool,
  onChange: React.PropTypes.func,
  onSelect: React.PropTypes.func,
  onDeselect: React.PropTypes.func
};

Select.defaultProps = {
  prefixCls: 'rc-select',
  filterOption: true,
  allowClear: true,
  onChange: noop,
  onSelect: noop,
  onDeselect: noop,
  notFoundContent: 'Not Found'
};

module.exports = Select;
