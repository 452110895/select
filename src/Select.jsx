'use strict';


import React from 'react';
import {joinClasses, classSet, KeyCode} from 'rc-util';
import anim from 'css-animation';
import OptGroup from './OptGroup';
import domAlign from 'dom-align';
import SelectDropdown from './Dropdown';
import {
  getPropValue, getValuePropValue, isCombobox,
  isMultipleOrTags, isMultipleOrTagsOrCombobox,
  isSingleMode, normValue
  } from './util';

function noop() {
}

function filterFn(input, child) {
  return getPropValue(child, this.props.optionFilterProp).indexOf(input) > -1;
}

function saveRef(name, component) {
  this[name] = component;
}

export default
class Select extends React.Component {
  constructor(props) {
    super(...arguments);
    var value = [];
    if ('value' in props) {
      value = normValue(props.value);
    } else if ('defaultValue' in props) {
      value = normValue(props.defaultValue);
    }
    this.state = {
      value: value,
      inputValue: ''
    };
    ['handleClick',
      'handleKeyDown',
      'handleInputKeyDown',
      'handleInputChange',
      'handleFocus',
      'handleBlur',
      'handleClearSelection',
      'handleMenuSelect',
      'handleMenuDeselect'].forEach((m)=> {
        this[m] = this[m].bind(this);
      });
    this.saveInputRef = saveRef.bind(this, 'inputInstance');
    this.saveDropdownRef = saveRef.bind(this, 'dropdownInstance');
  }

  componentWillReceiveProps(nextProps) {
    if ('value' in nextProps) {
      this.setState({
        value: normValue(nextProps.value)
      });
    }
  }

  fireChange(value) {
    this.props.onChange(isMultipleOrTags(this.props) ? value : value[0]);
    this.setState({
      value: value
    });
  }

  getLabelByValue(children, value) {
    if (value === undefined) {
      return null;
    }
    var label = null;
    React.Children.forEach(children, (c) => {
      if (c.type === OptGroup) {
        var maybe = this.getLabelByValue(c.props.children, value);
        if (maybe != null) {
          label = maybe;
        }
      } else if (getValuePropValue(c) === value) {
        label = getPropValue(c, this.props.optionLabelProp);
      }
    });
    return label;
  }

  setOpenState(open) {
    var refs = this.refs;
    this.setState({
      open: open
    }, ()=> {
      if (open || isMultipleOrTagsOrCombobox(this.props)) {
        if (this.getInputDOMNode()) {
          this.getInputDOMNode().focus();
        }
      } else if (refs.selection) {
        React.findDOMNode(refs.selection).focus();
      }
    });
  }

  handleInputChange(e) {
    var val = e.target.value;
    var props = this.props;
    this.setState({
      inputValue: val,
      open: true
    });
    if (isCombobox(props)) {
      props.onChange(val);
    }
    props.onSearch(val);
  }

  handleClick() {
    if (!this.props.disabled) {
      if (this.state.open) {
        this.setOpenState(false);
      } else {
        this.openIfHasChildren();
      }
    }
  }

  openIfHasChildren() {
    var props = this.props;
    if (React.Children.count(props.children) || isSingleMode(props)) {
      this.setOpenState(true);
    }
  }

  // combobox ignore
  handleKeyDown(e) {
    var props = this.props;
    if (props.disabled) {
      return;
    }
    var keyCode = e.keyCode;
    if (this.state.open && !this.getInputDOMNode()) {
      this.handleInputKeyDown(e);
    } else if (keyCode === KeyCode.ENTER || keyCode === KeyCode.DOWN) {
      this.handleClick();
      e.preventDefault();
    }
  }

  handleInputKeyDown(e) {
    var props = this.props;
    var state = this.state;
    var keyCode = e.keyCode;
    if (isMultipleOrTags(props) && !e.target.value && keyCode === KeyCode.BACKSPACE) {
      var value = state.value.concat();
      if (value.length) {
        value.pop();
        this.fireChange(value);
      }
      return;
    }

    if (keyCode === KeyCode.DOWN) {
      if (!state.open) {
        this.openIfHasChildren();
        e.preventDefault();
        e.stopPropagation();
        return;
      }
    } else if (keyCode === KeyCode.ESC) {
      if (state.open) {
        this.setOpenState(false);
        e.preventDefault();
        e.stopPropagation();
      }
      return;
    }

    if (state.open) {
      var menu = this.dropdownInstance && this.dropdownInstance.refs.menu;
      if (menu && menu.handleKeyDown(e)) {
        e.preventDefault();
        e.stopPropagation();
      }
    }
  }

  handleMenuSelect(key, item) {
    var value = this.state.value;
    var props = this.props;
    var selectedValue = getValuePropValue(item);
    if (value.indexOf(selectedValue) !== -1) {
      return;
    }
    if (isMultipleOrTags(props)) {
      value = value.concat([selectedValue]);
    } else {
      if (value[0] === selectedValue) {
        this.setOpenState(false);
        return;
      }
      value = [selectedValue];
    }
    props.onSelect(selectedValue, item);
    this.fireChange(value);
    this.setState({
      inputValue: ''
    });
    this.setOpenState(false);
    if (isCombobox(props)) {
      this.setState({
        inputValue: getPropValue(item, props.optionLabelProp)
      });
    }
  }

  handleMenuDeselect(key, item, e) {
    if (e.type === 'click') {
      this.removeSelected(getValuePropValue(item));
    }
    this.setState({
      inputValue: ''
    });
    this.setOpenState(false);
  }

  handleBlur() {
    if (this._blurTimer) {
      clearTimeout(this._blurTimer);
    }
    this._blurTimer = setTimeout(() => {
      this.setState({
        open: false
      });
    }, 100);
  }

  handleFocus() {
    if (this._blurTimer) {
      clearTimeout(this._blurTimer);
      this._blurTimer = null;
    }
  }

  removeSelected(selectedValue) {
    var props = this.props;
    if (props.disabled) {
      return;
    }
    var value = this.state.value.filter((v)=> {
      return v !== selectedValue;
    });
    var canMultiple = isMultipleOrTags(props);
    if (canMultiple) {
      props.onDeselect(selectedValue);
    }
    this.fireChange(value);
  }

  handleClearSelection(e) {
    var props = this.props;
    var state = this.state;
    if (props.disabled) {
      return;
    }
    e.stopPropagation();
    if (state.inputValue || state.value.length) {
      this.fireChange([]);
      this.setState({
        inputValue: ''
      });
    }
    this.setOpenState(false);
  }

  renderTopControlNode() {
    var value = this.state.value;
    var props = this.props;
    var prefixCls = props.prefixCls;
    var allowClear = props.allowClear;
    var children = props.children;
    var clear = <span className={prefixCls + '-selection__clear'}
      onClick={this.handleClearSelection}>Ã—</span>;
    // single and not combobox, input is inside dropdown
    if (isSingleMode(props)) {
      return (<span className={prefixCls + '-selection__rendered'}>
        <span>{this.getLabelByValue(children, value[0]) || props.placeholder}</span>
          {allowClear ? clear : null}
      </span>);
    } else {
      var selectedValueNodes;
      if (isMultipleOrTags(props)) {
        selectedValueNodes = value.map((v) => {
          var content = this.getLabelByValue(children, v) || v;
          var title = content;
          var maxTagTextLength = props.maxTagTextLength;
          if (maxTagTextLength && typeof content === 'string' && content.length > maxTagTextLength) {
            content = content.slice(0, maxTagTextLength) + '...';
          }
          return (
            <li className={prefixCls + '-selection__choice'}
              key={v}
              title={title}>
              <span className={prefixCls + '-selection__choice__content'}>{content}</span>
              <span className={prefixCls + '-selection__choice__remove'}
                onClick={this.removeSelected.bind(this, v)}/>
            </li>
          );
        });
      }
      return (
        <ul className={prefixCls + '-selection__rendered'}>
          {selectedValueNodes}
          {allowClear && !isMultipleOrTags(props) ? clear : null}
          <li className={joinClasses(prefixCls + '-search', prefixCls + '-search--inline')}>
          {this.getInputElement()}
          </li>
        </ul>
      );
    }
  }

  getDropdownDOMNode() {
    return React.findDOMNode(this.dropdownInstance);
  }

  getDropdownContainer() {
    if (!this.dropdownContainer) {
      this.dropdownContainer = document.createElement('div');
      document.body.appendChild(this.dropdownContainer);
    }
    return this.dropdownContainer;
  }

  renderDropdown(prevState) {
    var state = this.state;
    var props = this.props;
    var dropdownDOMNode;
    if (state.open && props.renderDropdownToBody) {
      React.render(this.getDropdownElement(), this.getDropdownContainer());
    }
    if (this.dropdownContainer) {
      this.dropdownContainer.className = `${this.props.prefixCls}-dropdown-container${state.open ? '-open' : ''}`;
    }
    if (props.dropdownMatchSelectWidth && state.open) {
      dropdownDOMNode = this.getDropdownDOMNode();
      if (dropdownDOMNode) {
        dropdownDOMNode.style.width = React.findDOMNode(this).offsetWidth + 'px';
      }
    }
    if (!prevState.open && state.open) {
      dropdownDOMNode = this.getDropdownDOMNode();
      if (dropdownDOMNode) {
        domAlign(dropdownDOMNode, React.findDOMNode(this), {
          points: ['tl', 'bl'],
          offset: [0, 4]
        });
      }
    }
  }

  getInputElement() {
    var props = this.props;
    return <input ref={this.saveInputRef}
      onChange={this.handleInputChange}
      onKeyDown={this.handleInputKeyDown}
      value={this.state.inputValue}
      disabled={props.disabled}
      placeholder={props.searchPlaceholder}
      className={props.prefixCls + '-search__field'}
      role="textbox" />;
  }

  getDropdownElement() {
    var state = this.state;
    var props = this.props;
    if (state.open) {
      this.cachedDropDown = <SelectDropdown
        key="dropdown"
        onDropdownFocus={this.handleFocus}
        onDropdownBlur={this.handleBlur}
        filterOption={props.filterOption}
        optionFilterProp={props.optionFilterProp}
        optionLabelProp={props.optionLabelProp}
        inputValue={state.inputValue}
        inputElement={this.getInputElement()}
        ref={this.saveDropdownRef}
        tags={props.tags}
        notFoundContent={props.notFoundContent}
        onMenuDeselect={this.handleMenuDeselect}
        onMenuSelect={this.handleMenuSelect}
        value={state.value}
        isMultipleOrTags={isMultipleOrTags(props)}
        prefixCls={props.prefixCls}
        isMultipleOrTagsOrCombobox={isMultipleOrTagsOrCombobox(props)}
        showSearch={props.showSearch}
        dropdownStyle={props.dropdownStyle}>
      {props.children}
      </SelectDropdown>;
    }
    return this.cachedDropDown;
  }

  animateDropdown(prevProps, prevState) {
    var props = this.props;
    var state = this.state;
    var transitionName = props.transitionName;
    if (!transitionName && props.animation) {
      transitionName = `${props.prefixCls}-dropdown-${props.animation}`;
    }
    var domNode = this.getDropdownDOMNode();
    if (transitionName && domNode) {
      if (state.open && !prevState.open) {
        anim(domNode, `${transitionName}-enter`);
      } else if (!state.open && prevState.open) {
        anim(domNode, `${transitionName}-leave`);
      }
    }
  }

  getInputDOMNode() {
    return React.findDOMNode(this.inputInstance);
  }

  componentDidMount() {
    this.componentDidUpdate();
  }

  componentDidUpdate(prevProps, prevState) {
    prevState = prevState || {};
    this.renderDropdown(prevState);
    this.animateDropdown(prevProps, prevState);
    if (isMultipleOrTags(this.props)) {
      var inputNode = this.getInputDOMNode();
      if (inputNode.value) {
        inputNode.style.width = inputNode.scrollWidth + 'px';
      } else {
        inputNode.style.width = '';
      }
    }
  }

  componentWillUnmount() {
    if (this.dropdownContainer) {
      React.unmountComponentAtNode(this.dropdownContainer);
      document.body.removeChild(this.dropdownContainer);
      this.dropdownContainer = null;
    }
    this.dropdownInstance = null;
    if (this._blurTimer) {
      clearTimeout(this._blurTimer);
      this._blurTimer = null;
    }
  }

  render() {
    var props = this.props;
    var multiple = isMultipleOrTags(props);
    var state = this.state;
    var prefixCls = props.prefixCls;
    var ctrlNode = this.renderTopControlNode();
    var extraSelectionProps = {};
    if (!isCombobox(props)) {
      extraSelectionProps = {
        onKeyDown: this.handleKeyDown,
        tabIndex: 0
      };
    }
    var rootCls = {
      [prefixCls]: true,
      [prefixCls + '-open']: this.state.open,
      [prefixCls + '-combobox']: isCombobox(props),
      [prefixCls + '-disabled']: props.disabled
    };
    return (
      <span
        style={props.style}
        className={joinClasses(props.className, classSet(rootCls))}
        onFocus={this.handleFocus}
        onBlur={this.handleBlur}>
        <span ref="selection"
          key="selection"
          className={joinClasses(prefixCls + '-selection',
            `${prefixCls}-selection--${multiple ? 'multiple' : 'single'}`)}
          role="combobox"
          aria-autocomplete="list"
          onClick={this.handleClick}
          aria-haspopup="true"
          aria-expanded={state.open}
      {...extraSelectionProps}
        >
        {ctrlNode}
        {multiple || !props.showArrow ? null :
          <span key="arrow" className={prefixCls + '-arrow'} tabIndex="-1" style={{outline: 'none'}}>
            <b></b>
          </span>}
        </span>
      {props.renderDropdownToBody ? null : this.getDropdownElement()}
      </span>
    );
  }
}

Select.propTypes = {
  multiple: React.PropTypes.bool,
  filterOption: React.PropTypes.any,
  showSearch: React.PropTypes.bool,
  showArrow: React.PropTypes.bool,
  renderDropdownToBody: React.PropTypes.bool,
  tags: React.PropTypes.bool,
  transitionName: React.PropTypes.string,
  optionLabelProp: React.PropTypes.string,
  optionFilterProp: React.PropTypes.string,
  animation: React.PropTypes.string,
  onChange: React.PropTypes.func,
  onSelect: React.PropTypes.func,
  onSearch: React.PropTypes.func,
  searchPlaceholder: React.PropTypes.string,
  placeholder: React.PropTypes.any,
  onDeselect: React.PropTypes.func,
  dropdownStyle: React.PropTypes.object,
  maxTagTextLength: React.PropTypes.number
};

Select.defaultProps = {
  prefixCls: 'rc-select',
  filterOption: filterFn,
  showSearch: true,
  allowClear: false,
  placeholder: '',
  searchPlaceholder: '',
  onChange: noop,
  onSelect: noop,
  onSearch: noop,
  onDeselect: noop,
  showArrow: true,
  dropdownMatchSelectWidth: true,
  dropdownStyle: {},
  renderDropdownToBody: false,
  optionFilterProp: 'value',
  optionLabelProp: 'value',
  notFoundContent: 'Not Found'
};
