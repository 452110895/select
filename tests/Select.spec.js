/* eslint-disable no-undef */
import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import TestUtils, { Simulate } from 'react-addons-test-utils';
import Select, { Option } from 'rc-select';
import $ from 'jquery';
import { mount, render } from 'enzyme';
import { renderToJson } from 'enzyme-to-json';

describe('Select', () => {
  let instance;
  let div;

  beforeEach(() => {
    div = document.createElement('div');
    document.body.appendChild(div);
  });

  afterEach(() => {
    ReactDOM.unmountComponentAtNode(div);
    document.body.removeChild(div);
  });

  it('render to body works', (done) => {
    instance = ReactDOM.render(
      <Select
        value="2"
      >
        <Option value="1">1</Option>
        <Option value="2">2</Option>
      </Select>,
      div);
    instance.setState({
      open: true,
    }, () => {
      expect(instance.getPopupDOMNode().parentNode
        .parentNode.parentNode.nodeName.toLowerCase()).toBe('body');
      expect(instance.getPopupDOMNode().className).not.toContain('hidden');
      done();
    });
  });

  it('renders correctly', () => {
    const wrapper = render(
      <Select className="forTest" openClassName="my-open" value="2">
        <Option value="1">1</Option>
        <Option value="2" disabled>2</Option>
      </Select>
    );

    expect(renderToJson(wrapper)).toMatchSnapshot();
  });

  it('should default select the right option', (done) => {
    instance = ReactDOM.render(
      <Select defaultValue="2">
        <Option value="1">1</Option>
        <Option value="2">2</Option>
      </Select>, div);
    instance.setState({
      open: true,
    }, () => {
      expect(instance.getPopupMenuComponent().instanceArray[0].isSelected()).toBe(false);
      expect(instance.getPopupMenuComponent().instanceArray[1].isSelected()).toBe(true);
      done();
    });
  });

  it('should can select multiple items', (done) => {
    instance = ReactDOM.render(
      <Select multiple value={['1', '2']}>
        <Option value="1">1</Option>
        <Option value="2">2</Option>
        <Option value="3">2</Option>
      </Select>, div);
    instance.setState({
      open: true,
    }, () => {
      expect(instance.getPopupMenuComponent().instanceArray[0].isSelected()).toBe(true);
      expect(instance.getPopupMenuComponent().instanceArray[1].isSelected()).toBe(true);
      expect(instance.getPopupMenuComponent().instanceArray[2].isSelected()).toBe(false);
      done();
    });
  });

  it('should show clear button', () => {
    instance = ReactDOM.render(
      <Select value="1" allowClear>
        <Option value="1">1</Option>
        <Option value="2">2</Option>
      </Select>,
      div);
    expect(TestUtils.scryRenderedDOMComponentsWithClass(instance,
      'rc-select-selection__clear')[0].style.display).toBe('block');
  });

  it('should hide clear button', () => {
    instance = ReactDOM.render(
      <Select allowClear>
        <Option value="1">1</Option>
        <Option value="2">2</Option>
      </Select>,
      div);
    expect(TestUtils.scryRenderedDOMComponentsWithClass(instance,
      'rc-select-selection__clear')[0].style.display).toBe('none');
  });

  it('should not response click event when select is disabled', (done) => {
    instance = ReactDOM.render(
      <Select disabled defaultValue="2">
        <Option value="1">1</Option>
        <Option value="2">2</Option>
      </Select>, div);
    Simulate.click(ReactDOM.findDOMNode(instance.refs.selection));
    expect(instance.state.open).not.toBeTruthy();
    done();
  });

  it('should show selected value in singleMode when close', (done) => {
    instance = ReactDOM.render(
      <Select value="1">
        <Option value="1">1</Option>
        <Option value="2">2</Option>
      </Select>, div);
    expect($(ReactDOM.findDOMNode(instance))
      .find('.rc-select-selection-selected-value').length).toBe(1);
    done();
  });

  it('should show placeholder in singleMode when value is undefined', (done) => {
    instance = ReactDOM.render(
      <Select placeholder="aaa">
        <Option value="1">1</Option>
        <Option value="2">2</Option>
      </Select>, div);
    expect($(ReactDOM.findDOMNode(instance))
      .find('.rc-select-selection__placeholder').length).toBe(1);
    done();
  });

  describe('when open', () => {
    // this.timeout(400000);

    beforeEach((done) => {
      div = document.createElement('div');

      div.tabIndex = 0;
      document.body.appendChild(div);

      instance = ReactDOM.render(
        <Select>
          <Option value="1">1</Option>
          <Option value="2">2</Option>
        </Select>,
        div);


      Simulate.click(ReactDOM.findDOMNode(instance.refs.selection));
      done();
    });

    afterEach(() => {
      ReactDOM.unmountComponentAtNode(div);
    });

    it('should show not found', (done) => {
      instance.getInputDOMNode().value = '4';
      Simulate.change(instance.getInputDOMNode());
      setTimeout(() => {
        expect($(instance.getPopupDOMNode())
          .find('.rc-select-dropdown-menu-item').length).toBe(1);
        expect($(instance.getPopupDOMNode())
          .find('.rc-select-dropdown-menu-item')[0].innerHTML).toBe('Not Found');
        done();
      }, 100);
    });

    it('should show search input in single selection trigger', (done) => {
      expect($(instance.getInputDOMNode()).parents('.rc-select-open').length).toBe(1);
      done();
    });
  });

  describe('automatic tokenization ', () => {
    it('tokenize tag select', () => {
      const wrapper = mount(
        <Select tags tokenSeparators={[',']}>
          <Option value="1">1</Option>
          <Option value="2">2</Option>
        </Select>,
        div);

      wrapper.find('input').simulate('change', { target: {
        value: '2,3,4',
      } });

      expect(wrapper.state().value).toEqual([
        { key: '2', label: '2' },
        { key: '3', label: '3' },
        { key: '4', label: '4' },
      ]);
    });

    it('tokenize multiple select', () => {
      const wrapper = mount(
        <Select multiple optionLabelProp="children" tokenSeparators={[',']}>
          <Option value="1">One</Option>
          <Option value="2">Two</Option>
        </Select>,
      );

      const input = wrapper.find('input');

      input.simulate('change', { target: {
        value: 'One',
      } });

      input.simulate('change', { target: {
        value: 'One,Two,Three',
      } });

      expect(wrapper.state().value).toEqual([
        { key: '1', label: 'One' },
        { key: '2', label: 'Two' },
      ]);
    });
  });

  it('display correct combobox label when it\'s under controllered', (done) => {
    class App extends Component {
      state = {
        value: {
          label: '',
          key: '',
        },
      }

      render() {
        return (
          <Select
            combobox
            labelInValue
            value={this.state.value}
            optionLabelProp="children"
          >
            <Option value="1">One</Option>
            <Option value="2">Two</Option>
          </Select>
        );
      }
    }

    const wrapper = mount(<App />);

    wrapper.setState({ value: { label: 'One', key: '1' } }, () => {
      expect(wrapper.find('input').props().value).toBe('One');
      done();
    });
  });
});
