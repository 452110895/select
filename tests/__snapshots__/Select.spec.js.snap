exports[`Select renders correctly 1`] = `
<div
  class="forTest rc-select rc-select-enabled">
  <div
    aria-autocomplete="list"
    aria-expanded="false"
    aria-haspopup="true"
    class="rc-select-selection
              rc-select-selection--single"
    role="combobox"
    tabindex="0">
    <div
      class="rc-select-selection__rendered">
      <div
        class="rc-select-selection-selected-value"
        style="display:block;opacity:1;"
        title="2">
        2
      </div>
      <div
        class="rc-select-search rc-select-search--inline"
        style="display:none;">
        <div
          class="rc-select-search__field__wrap">
          <input
            class="rc-select-search__field"
            value="" />
          <span
            class="rc-select-search__field__mirror" />
        </div>
      </div>
    </div>
    <span
      class="rc-select-arrow"
      style="user-select:none;-webkit-user-select:none;"
      unselectable="unselectable">
      <b />
    </span>
  </div>
</div>
`;
